### PR TITLE
Add categorical color mapper

### LINKED ISSUE #5013: Discrete/Categorical ColorMapper and ColorBar
Add a CategoricalColorMapper that accepts palette and factors

```python
color_mapper = CategoricalColorMapper(
   palette=Spectral3,
   factors=["cat", "dog", "milk"],
)
```

@canavandl - we'd want color bar to use a CategoricalAxis and I'm not sure we could use image - maybe we could though. It's very similar to low-n palettes.

@bryevdv - `nan_color`, `high_color`, `low_color`. high/low don't make sense but out_of_range does. Which I see as different from No Data. Thoughts on naming.

### PR DESCRIPTION
**Pull request by birdsarah**
Fixes #5013

### CODE REVIEW

File: bokeh/models/mappers.py [hunk 1]
```diff
@@ -80,3 +92,13 @@ class LogColorMapper(ColorMapper):
...
+
+    low = Float(help="""
+    The minimum value of the range to map into the palette. Values below
+    this are clamped to ``low``.
+    """)
+
+    high = Float(help="""
+    The maximum value of the range to map into the palette. Values above
+    this are clamped to ``high``.
+    """)
```
Comment (bryevdv) [inline bokeh/models/mappers.py]: We should probably factor out a `ContinuousColorMapper` for linear and log mappers to inherit from.

**Comment by canavandl** [comment]
The categorical color mapper won't work with a ColorBar (which is fine, a `Legend` is probably a better annotation here anyway), because it lacks high/low attributes.

We should probably add a `ContinuousColorMapper` base class that ColorBar accepts (as per @bryevdv) to limit confusion.

**Comment by birdsarah** [comment]
Split out the ContinuousColorMapper.

File: bokeh/models/mappers.py [hunk 2]
```diff
@@ -24,6 +24,35 @@ class ColorMapper(Model):
...
+        super(ColorMapper, self).__init__(**kwargs)
+
+
+class CategoricalColorMapper(ColorMapper):
+    """ Map categories to colors. Values that are passed to
+    this mapper that aren't in factors will be assigned the nan_color.
+
+    """
+
+    factors = Either(List(String), List(Int), List(Float), List(Datetime), List(Date), help="""
```
Comment (bryevdv) [inline bokeh/models/mappers.py]: `Seq` ? Is there a reason it has to be a list and not a tuple or an array?

Comment (birdsarah) [inline bokeh/models/mappers.py]: Doh! Thanks.

**Comment by bryevdv** [comment]
LGTM I think we should update all examples and notebooks to use this, but I'll make a separate issue for that. Maybe we can also add some slight convenience syntax around these things at that time too.

File: bokeh/models/mappers.py (hunk 2 above)
Comment (bryevdv) [inline bokeh/models/mappers.py]: Crap, I was in a hurry and forgot about this before merging. I'll revert it.

**Pull request merged**